### ISSUE TITLE
[2.1.0] module:uninstall can remove code it uses itself

### ISSUE THREAD

**Issue by hostep** [magento/magento2#5797]
### Preconditions
1. Figure out a way to travel back and forth in time (I added some _hints_)
2. We use `composer` version 1.1.3 by adding this to our composer.json file: `"composer/composer": "1.1.3 as 1.0.0-beta1",`, due to [another bug](https://github.com/magento/magento2/issues/4359#issuecomment-225115951), but I think it won't really matter which version is being used
3. PHP 7.0.9
4. Magento 2.1.0
### Steps to reproduce

1. Travel back in time to when `symfony/process` version 2.8.8 wasn't released yet (_hint: try to downgrade to version 2.8.7 of the module after step 2_)
2. Install Magento 2.1.0 using composer
3. Install a third party module using composer (`composer require`, `bin/magento setup:upgrade`, ...)
4. Travel back to the current time when `symfony/process` version 2.8.8 was released (_hint: don't do anything in this step_)
5. Uninstall the module you installed in step 3 by using `bin/magento module:uninstall`
### Expected result
1. Module is uninstalled without a problem
### Actual result
1. When Magento tries to uninstall the module using `composer`, it also at the same time tries to update `symfony/process` from version 2.8.7 to 2.8.8 for some strange reason. And since `symfony/process` is being used by the `module:uninstall` command, the command stops running.
2. Console output:

```
➜ php bin/magento module:uninstall Vendor_MyModule
You are about to remove code and/or database tables. Are you sure?[y/N]y
Enabling maintenance mode
You are about to remove a module(s) that might have database data. Do you want to remove the data from database?[y/N]N
Removing Vendor_MyModule from module registry in database
Removing Vendor_MyModule from module list in deployment configuration
Removing code from Magento codebase:
Command "remove" failed: Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing vendor/my-module (v1.0.11)
Removing Vendor/MyModule
  - Removing symfony/process (v2.8.7)

Removal failed, reverting ./composer.json to its original content.



  [ErrorException]
  include(/Users/me/Projects/my-project/vendor/composer/../symfony/process/ExecutableFinder.php): failed to open stream: No such file or directory



remove [--dev] [--no-progress] [--no-update] [--update-no-dev] [--update-with-dependencies] [--no-update-with-dependencies] [--ignore-platform-reqs] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [packages1] ... [packagesN]



Please disable maintenance mode after you resolved above issues
```
### Bit of debugging

It turns out at the step when the `composer remove` step is executed, this happens:

```
➜ composer remove --working-dir=/Users/me/Projects/my-project vendor/my-module
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing vendor/my-module (v1.0.11)
Removing Vendor/MyModule
  - Removing symfony/process (v2.8.7)
  - Installing symfony/process (v2.8.8)
    Loading from cache

Package fabpot/php-cs-fixer is abandoned, you should avoid using it. Use friendsofphp/php-cs-fixer instead.
Writing lock file
Generating optimized autoload files
```

So for some reason `symfony/process` gets updated to the latest version while removing another module which has no direct dependencies on `symfony/process`. But this particular module has some dependencies on some magento modules which in itself will probably have dependencies on `symfony/process` module.

I think you can work around this if you add the `--no-update-with-dependencies` switch, since that won't update `symfony/process`:

```
➜ composer remove --no-update-with-dependencies --working-dir=/Users/me/Projects/my-project vendor/my-module
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing vendor/my-module (v1.0.11)
Removing Vendor/MyModule
Package fabpot/php-cs-fixer is abandoned, you should avoid using it. Use friendsofphp/php-cs-fixer instead.
Writing lock file
Generating optimized autoload files
```

Please review thoroughly, because this seems to be something which can turn out to be quite complicated!
Also: I haven't tested my solution with that `--no-update-with-dependencies` flag yet.

For other people running into this problem: first update all your dependencies to the latest version by just running `composer update` (_hint_ you can run `composer update --dry-run` to preview what it will do first) and only after your dependencies are updated, run the `module:uninstall` command.

**Comment by hostep** [comment]
Possible related issue: https://github.com/magento/magento2/issues/3544

**Comment by shiftedreality** [comment]
Hi @hostep 

Thank you for deep investigation and reporting.
We've create internal ticket MAGETWO-58132 to fix this issue.

**Comment by mattcoz** [comment]
I'm unable to upgrade from 2.1.3 to 2.1.6 because of this error. It removes symfony/process and then immediately throws the error. It happens during the "composer update" command, so that doesn't help me.

**Comment by magento-engcom-team** [comment]
@hostep, thank you for your report.
The issue is already fixed in 2.2.0

**Comment by sidolov** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#18002 by @Thundar in 2.2-develop branch
Related commit(s):
 - [d5b370a2c86476fce16df71207cf05e63dac4ac1](https://github.com/magento/magento2/commit/d5b370a2c86476fce16df71207cf05e63dac4ac1)
 - [bf9b7a89ea3030209e83c3849835ccdac14aa0ab](https://github.com/magento/magento2/commit/bf9b7a89ea3030209e83c3849835ccdac14aa0ab)

The fix will be available with the upcoming 2.2.8 release.

**Comment by sidolov** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#18205 by @mage2pratik in 2.3-develop branch
Related commit(s):
 - [8b9fff56f2172bb2e3fa5bde4d16daa18df97e22](https://github.com/magento/magento2/commit/8b9fff56f2172bb2e3fa5bde4d16daa18df97e22)
 - [fa4289fe8534a2af5aeb69839ba0df6aa21a55ed](https://github.com/magento/magento2/commit/fa4289fe8534a2af5aeb69839ba0df6aa21a55ed)

The fix will be available with the upcoming 2.3.1 release.

**Comment by Eddcapone** [comment]
I have the same issue in Magento 2.4.2

**Comment by hostep** [comment]
@Eddcapone: please open a new issue, the fix mentioned above is still in Magento 2.4.2 so you are probably running against a new bug.

**Comment by haijerome** [comment]
apparently I ran **`composer remove vendor/module --ignore-platform-reqs`** to uninstall the module that I had installed through composer.
